### ISSUE TITLE
The gradientTest error of modification Euclidean loss. Need Help~!!!

### ISSUE THREAD

**Issue by winboyer** [BVLC/caffe#4430]
@Yangqing 
I'm a caffe beginner.
Because I need to train a regression model, I modified the Euclidean loss layer like below:

**The definition of hpp file:**
```
template <typename Dtype>  
class JointEucliLossLayer : public LossLayer<Dtype> {  
  public:
    explicit JointEucliLossLayer(const LayerParameter& param)
      : LossLayer<Dtype>(param), diff_() {}
    //virtual void LayerSetUp(const vector<Blob<Dtype>*>& bottom,  
    //  const vector<Blob<Dtype>*>& top);  
    virtual void Reshape(const vector<Blob<Dtype>*>& bottom,
      const vector<Blob<Dtype>*>& top);
    virtual inline const char* type() const { return "JointEucliLoss"; }  
    virtual inline bool AllowForceBackward(const int bottom_index) const {
      return true;
    }
   protected: 
     virtual void Forward_cpu(const vector<Blob<Dtype>*>& bottom,
      const vector<Blob<Dtype>*>& top);
     //virtual void Forward_gpu(const vector<Blob<Dtype>*>& bottom,
     // const vector<Blob<Dtype>*>& top);
    virtual void Backward_cpu(const vector<Blob<Dtype>*>& top,
      const vector<bool>& propagate_down, const vector<Blob<Dtype>*>& bottom);
    //virtual void Backward_gpu(const vector<Blob<Dtype>*>& top,
    //  const vector<bool>& propagate_down, const vector<Blob<Dtype>*>& bottom);
    Dtype alpha_box_;
    Dtype alpha_landmark_;
    Dtype beta_box_;
    Dtype beta_landmark_;
    Blob<Dtype> diff_;
    Blob<Dtype> target_labels_;
};
```

**The implement of cpp file:**
```
template <typename Dtype>
void JointEucliLossLayer<Dtype>::Reshape(
  const vector<Blob<Dtype>*>& bottom, const vector<Blob<Dtype>*>& top) {
  LossLayer<Dtype>::Reshape(bottom, top);
  CHECK_EQ(bottom[0]->count(1), bottom[1]->count(1)*4/6)
      << "Inputs must have the same dimension.";
  diff_.ReshapeLike(*bottom[0]);
  target_labels_.Reshape(bottom[1]->num(), bottom[1]->channels()*4/6, bottom[1]->height(), bottom[1]->width());
}
template <typename Dtype>
void JointEucliLossLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
    const vector<Blob<Dtype>*>& top) {
  //const int count = bottom[0]->count();
  const int num = bottom[0]->num();
  const int channels = bottom[0]->channels();
  const int channels_bt1 = bottom[1]->channels();
  int number = 0;
  const Dtype alpha_box = this->layer_param_.joint_eucliloss_param().alpha_box();
  for(int i=0;i<num;++i){
    for(int j=0;j<channels;++j){
      target_labels_.mutable_cpu_data()[i*channels+j] = bottom[1]->cpu_data()[i*channels_bt1+2+j];
    }
  }
  Dtype dot = 0.0;
  for(int i=0;i<num;++i){
    if(bottom[1]->cpu_data()[channels_bt1*i] != 0){
      for(int j=0;j<channels;++j){
        diff_.mutable_cpu_data()[i*channels+j] = bottom[0]->cpu_data()[i*channels+j]-target_labels_.cpu_data()[i*channels+j];
        number++;
        dot += diff_.cpu_data()[i*channels+j] * diff_.cpu_data()[i*channels+j];
      }
    }

  }
  //caffe_sub(count, bottom[0]->cpu_data(), target_labels_.cpu_data(), diff_.mutable_cpu_data());
  //caffe_sub(count, bottom[0]->cpu_data(), bottom[1]->cpu_data(), diff_.mutable_cpu_data());
  //Dtype dot = caffe_cpu_dot(count, diff_.cpu_data(), diff_.cpu_data());
  //Dtype loss = dot / bottom[0]->num() / Dtype(2) * alpha_box;
  Dtype loss = dot / (number) / Dtype(2) * alpha_box;
  top[0]->mutable_cpu_data()[0] = loss;
}
template <typename Dtype>
void JointEucliLossLayer<Dtype>::Backward_cpu(const vector<Blob<Dtype>*>& top,
    const vector<bool>& propagate_down, const vector<Blob<Dtype>*>& bottom) {
  const int count = bottom[0]->count();
  const int num = bottom[0]->num();
  const int channels = bottom[0]->channels();
  const int channels_bt1 = bottom[1]->channels();
  int number = 0;
  for(int i=0;i<num;++i){
    if(bottom[1]->cpu_data()[i*channels_bt1] != 0)
      number++;
    for(int j=0;j<channels;++j){
      target_labels_.mutable_cpu_data()[i*channels+j] = bottom[1]->cpu_data()[i*channels_bt1+2+j];
    }
  }
  const Dtype alpha_box = this->layer_param_.joint_eucliloss_param().alpha_box();
  if(propagate_down[0]){
    //const Dtype sign = (i == 0) ? 1 : -1;
    const Dtype sign = 1;
    //const Dtype alpha = sign * top[0]->cpu_diff()[0] / bottom[0]->num();
    const Dtype alpha = sign * top[0]->cpu_diff()[0] / (number);
    
    //caffe_cpu_axpby(bottom[i]->count(), alpha, diff_.cpu_data(), Dtype(0), bottom[i]->mutable_cpu_diff());
    for(int j=0;j<num;++j){
      if(bottom[1]->cpu_data()[channels_bt1*j] != 0){
        for(int k=0;k<channels;++k){
          bottom[0]->mutable_cpu_diff()[j*channels+k] = alpha*diff_.cpu_data()[j*channels+k];
        }
      }
    }
  }
  if(propagate_down[1]){
    const Dtype sign = -1;
    //const Dtype alpha = sign * top[0]->cpu_diff()[0] / bottom[0]->num();
    const Dtype alpha = sign * top[0]->cpu_diff()[0] / (number);
    Dtype* bottom_diff = bottom[0]->mutable_cpu_diff();
    //caffe_cpu_axpby(bottom[i]->count(), alpha, diff_.cpu_data(), Dtype(0), bottom[i]->mutable_cpu_diff());
    for(int j=0;j<num;++j){
      if(bottom[1]->cpu_data()[channels_bt1*j] != 0){
        for(int k=0;k<channels;++k){
          bottom[1]->mutable_cpu_diff()[j*channels+k+2] = alpha*diff_.cpu_data()[j*channels+k];
        }
      }
    }
  }
}
```

**Then, I test the forward loss and gradient like below:**
```
template <typename TypeParam>
class JointEucliLossLayerTest : public MultiDeviceTest<TypeParam> {
  typedef typename TypeParam::Dtype Dtype;

 protected:
  JointEucliLossLayerTest()
      : blob_bottom_data_(new Blob<Dtype>(10, 4, 1, 1)),
        blob_bottom_label_(new Blob<Dtype>(10, 6, 1, 1)),
        blob_top_loss_(new Blob<Dtype>()) {
    // fill the values
    FillBottomLabels();
    blob_bottom_vec_.push_back(blob_bottom_data_);
    blob_bottom_vec_.push_back(blob_bottom_label_);
    blob_top_vec_.push_back(blob_top_loss_);
  }
  
  virtual void FillBottomLabels() {
    FillerParameter filler_param;
    GaussianFiller<Dtype> filler(filler_param);
    filler.Fill(this->blob_bottom_data_);
    filler.Fill(this->blob_bottom_label_);
    const unsigned int prefetch_rng_seed = caffe_rng_rand();
    shared_ptr<Caffe::RNG> rng(new Caffe::RNG(prefetch_rng_seed));
    caffe::rng_t* prefetch_rng = static_cast<caffe::rng_t*>(rng->generator());
    Dtype* label_data = blob_bottom_label_->mutable_cpu_data();
    int channels = blob_bottom_label_->channels();
    for (int i = 0; i < blob_bottom_label_->num(); ++i) {
      label_data[i*channels] = (*prefetch_rng)() % 3;
      if(label_data[i*channels]==0)
        label_data[i*channels+1] = 1;
      else if(label_data[i*channels]==1)
        label_data[i*channels+1] = 0;
      else
        label_data[i*channels+1] = 2;
    }
  }
  
  virtual ~JointEucliLossLayerTest() {
    delete blob_bottom_data_;
    delete blob_bottom_label_;
    delete blob_top_loss_;
  }
  void TestForward() {
    // Get the loss without a specified objective weight -- should be
    // equivalent to explicitly specifiying a weight of 1.
    LayerParameter layer_param;
    JointEucliLossLayer<Dtype> layer_weight_1(layer_param);
    layer_weight_1.SetUp(this->blob_bottom_vec_, this->blob_top_vec_);
    const Dtype loss_weight_1 =
        layer_weight_1.Forward(this->blob_bottom_vec_, this->blob_top_vec_);

    // Get the loss again with a different objective weight; check that it is
    // scaled appropriately.
    const Dtype kLossWeight = 3.7;
    layer_param.add_loss_weight(kLossWeight);
    JointEucliLossLayer<Dtype> layer_weight_2(layer_param);
    layer_weight_2.SetUp(this->blob_bottom_vec_, this->blob_top_vec_);
    const Dtype loss_weight_2 =
        layer_weight_2.Forward(this->blob_bottom_vec_, this->blob_top_vec_);
    const Dtype kErrorMargin = 1e-5;
    EXPECT_NEAR(loss_weight_1 * kLossWeight, loss_weight_2, kErrorMargin);
    // Make sure the loss is non-trivial.
    const Dtype kNonTrivialAbsThresh = 1e-1;
    EXPECT_GE(fabs(loss_weight_1), kNonTrivialAbsThresh);
  }

  Blob<Dtype>* const blob_bottom_data_;
  Blob<Dtype>* const blob_bottom_label_;
  Blob<Dtype>* const blob_top_loss_;
  vector<Blob<Dtype>*> blob_bottom_vec_;
  vector<Blob<Dtype>*> blob_top_vec_;

};

TYPED_TEST_CASE(JointEucliLossLayerTest, TestDtypesAndDevices);

TYPED_TEST(JointEucliLossLayerTest, TestForward) {
  this->TestForward();
}

TYPED_TEST(JointEucliLossLayerTest, TestGradient) {
  typedef typename TypeParam::Dtype Dtype;
  LayerParameter layer_param;
  const Dtype kLossWeight = 3.7*0.5;
  layer_param.add_loss_weight(kLossWeight);
  JointEucliLossLayer<Dtype> layer(layer_param);
  layer.SetUp(this->blob_bottom_vec_, this->blob_top_vec_);
  GradientChecker<Dtype> checker(1e-2, 1e-2, 1701);
  checker.CheckGradientExhaustive(&layer, this->blob_bottom_vec_, this->blob_top_vec_);
}
```

**But I get the gradient test error like this**
```
debug: (top_id, top_data_id, blob_id, feat_id)=0,0,1,2; feat = -0.99483954906463623; objective+ = 1.3239332437515259; objective- = 1.3235070705413818
./include/caffe/test/test_gradient_check_util.hpp:175: Failure
The difference between computed_gradient and estimated_gradient is 0.086927413940429688, which exceeds threshold_ * scale, where
computed_gradient evaluates to 0,
estimated_gradient evaluates to -0.086927413940429688, and
threshold_ * scale evaluates to 0.0099999997764825821.
debug: (top_id, top_data_id, blob_id, feat_id)=0,0,1,3; feat = -0.12848135828971863; objective+ = 1.3228509426116943; objective- = 1.3245894908905029
./include/caffe/test/test_gradient_check_util.hpp:175: Failure
The difference between computed_gradient and estimated_gradient is 0.091773271560668945, which exceeds threshold_ * scale, where
computed_gradient evaluates to 0,
estimated_gradient evaluates to 0.091773271560668945, and
threshold_ * scale evaluates to 0.0099999997764825821.
```

Could anyone can help me?!!!

**Comment by cleardusk** [comment]
Did you solve it? I had a similar situation -_-
